src/Native/System.Security.Cryptography.Native/pal_rsa.cpp

```diff
@@ -12,3 +31,118 @@ extern "C" RSA* DecodeRsaPublicKey(const unsigned char* buf, int32_t len)
+{
+    int padding = useOaepPadding ? RSA_PKCS1_OAEP_PADDING : RSA_PKCS1_PADDING;
+    return RSA_private_decrypt(flen, from, to, rsa, padding);
+}
+
+extern "C" int32_t RsaSize(RSA* rsa)
+{
+    return RSA_size(rsa);
+}
+
+extern "C" int32_t RsaGenerateKeyEx(RSA* rsa, int32_t bits, BIGNUM* e)
+{
+    return RSA_generate_key_ex(rsa, bits, e, nullptr);
+}
+
+extern "C" int32_t RsaSign(int32_t type, const uint8_t* m, int32_t m_len, uint8_t* sigret, int32_t* siglen, RSA* rsa)
+{
+    if (!siglen)
+    {
+        return 0;
+    }
+
+    unsigned int unsignedSigLen = 0;
+    int32_t ret = RSA_sign(type, m, UnsignedCast(m_len), sigret, &unsignedSigLen, rsa);
+    assert(unsignedSigLen <= INT32_MAX);
+    *siglen = static_cast<int32_t>(unsignedSigLen);
+    return ret;
+}
+
+extern "C" int32_t RsaVerify(int32_t type, const uint8_t* m, int32_t m_len, uint8_t* sigbuf, int32_t siglen, RSA* rsa)
+{
+    return RSA_verify(type, m, UnsignedCast(m_len), sigbuf, UnsignedCast(siglen), rsa);
+}
+
+extern "C" void GetRsaParameters(const RSA* rsa,
+                                 BIGNUM** n,
+                                 BIGNUM** e,
+                                 BIGNUM** d,
+                                 BIGNUM** p,
+                                 BIGNUM** dmp1,
+                                 BIGNUM** q,
+                                 BIGNUM** dmq1,
+                                 BIGNUM** iqmp)
+{
+    if (!rsa || !n || !e || !d || !p || !dmp1 || !q || !dmq1 || !iqmp)
+    {
+        return;
+    }
+
+    *n = rsa->n;
+    *e = rsa->e;
+    *d = rsa->d;
+    *p = rsa->p;
+    *dmp1 = rsa->dmp1;
+    *q = rsa->q;
+    *dmq1 = rsa->dmq1;
+    *iqmp = rsa->iqmp;
+}
+
+static void SetRsaParameter(BIGNUM** rsaFieldAddress, uint8_t* buffer, int32_t bufferLength)
+{
+    if (rsaFieldAddress)
+    {
+        if (!buffer || !bufferLength)
+        {
+            *rsaFieldAddress = nullptr;
+        }
+        else
+        {
+            BIGNUM* bigNum = BN_bin2bn(buffer, bufferLength, nullptr);
+            *rsaFieldAddress = bigNum;
+        }
+    }
+}
+
+extern "C" void SetRsaParameters(RSA* rsa,
+                                 uint8_t* n,
+                                 int32_t nLength,
+                                 uint8_t* e,
+                                 int32_t eLength,
+                                 uint8_t* d,
+                                 int32_t dLength,
+                                 uint8_t* p,
+                                 int32_t pLength,
+                                 uint8_t* dmp1,
+                                 int32_t dmp1Length,
+                                 uint8_t* q,
+                                 int32_t qLength,
+                                 uint8_t* dmq1,
+                                 int32_t dmq1Length,
+                                 uint8_t* iqmp,
+                                 int32_t iqmpLength)
+{
+    if (!rsa)
+    {
+        return;
```

Comment:
I'll add the asserts. We never expect this to no-op. I was just being defensive with these checks.